pilot/pkg/networking/plugin/authn/authentication_test.go

```diff
@@ -534,3 +535,222 @@ func TestBuildJwtFilter(t *testing.T) {
+				}},
+			},
+			expected: &authn_filter.FilterConfig{
+				Policy: &authn.Policy{
+					Peers: []*authn.PeerAuthenticationMethod{{
+						Params: &authn.PeerAuthenticationMethod_Mtls{
+							&authn.MutualTls{},
+						},
+					}},
+				},
+			},
+		},
+		{
+			name: "jwt policy",
+			in: &authn.Policy{
+				Peers: []*authn.PeerAuthenticationMethod{
+					{
+						Params: &authn.PeerAuthenticationMethod_Jwt{
+							Jwt: &authn.Jwt{
+								Issuer: "foo",
+							},
+						},
+					},
+				},
+			},
+			expected: &authn_filter.FilterConfig{
+				Policy: &authn.Policy{
+					Peers: []*authn.PeerAuthenticationMethod{
+						{
+							Params: &authn.PeerAuthenticationMethod_Jwt{
+								Jwt: &authn.Jwt{
+									Issuer: "foo",
+								},
+							},
+						},
+					},
+				},
+				JwtOutputPayloadLocations: map[string]string{
+					"foo": "istio-sec-0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33",
+				},
+			},
+		},
+		{
+			name: "complex",
+			in: &authn.Policy{
+				Targets: []*authn.TargetSelector{
+					{
+						Name: "svc1",
+					},
+				},
+				Peers: []*authn.PeerAuthenticationMethod{
+					{
+						Params: &authn.PeerAuthenticationMethod_Jwt{
+							Jwt: &authn.Jwt{
+								Issuer: "foo",
+							},
+						},
+					},
+					{
+						Params: &authn.PeerAuthenticationMethod_Mtls{},
+					},
+				},
+				Origins: []*authn.OriginAuthenticationMethod{
+					{
+						Jwt: &authn.Jwt{
+							Issuer: "bar",
+						},
+					},
+				},
+			},
+			expected: &authn_filter.FilterConfig{
+				Policy: &authn.Policy{
+					Peers: []*authn.PeerAuthenticationMethod{
+						{
+							Params: &authn.PeerAuthenticationMethod_Jwt{
+								Jwt: &authn.Jwt{
+									Issuer: "foo",
+								},
+							},
+						},
+						{
+							Params: &authn.PeerAuthenticationMethod_Mtls{
+								&authn.MutualTls{},
+							},
+						},
+					},
+					Origins: []*authn.OriginAuthenticationMethod{
+						{
+							Jwt: &authn.Jwt{
+								Issuer: "bar",
+							},
+						},
+					},
+				},
+				JwtOutputPayloadLocations: map[string]string{
+					"foo": "istio-sec-0beec7b5ea3f0fdbc95d0dd47f3c5bc275da8a33",
+					"bar": "istio-sec-62cdb7020ff920e5aa642c3d4066950dd1f01f4d",
+				},
+			},
+		},
+	}
+	for _, c := range cases {
+		if got := ConvertPolicyToAuthNFilterConfig(c.in); !reflect.DeepEqual(c.expected, got) {
+			t.Errorf("Test case %s: expected\n%#v\n, got\n%#v", c.name, c.expected.String(), got.String())
+		}
+	}
+}
+
+func TestBuildAuthNFilter(t *testing.T) {
+	cases := []struct {
+		in       *authn.Policy
+		expected *http_conn.HttpFilter
+	}{
+		{
+			in:       nil,
+			expected: nil,
+		},
+		{
+			in:       &authn.Policy{},
+			expected: nil,
+		},
+		{
+			in: &authn.Policy{
+				Peers: []*authn.PeerAuthenticationMethod{
+					{
+						Params: &authn.PeerAuthenticationMethod_Jwt{
+							Jwt: &authn.Jwt{
+								JwksUri: "http://abc.com",
+							},
+						},
+					},
+				},
+			},
+			expected: &http_conn.HttpFilter{
```

Comment:
Can you clean this up, e.g. by using the original type inline and wrapping it in a call to convert it to a struct? Struct protos are really hard to read.

Reply:
I change the test to convert config back to proto and compare with golden proto value. Hope it's easier to read

(will clean up other test in another if this is a reasonable way to do)